doc/design/high_level_design_document.md

```diff
@@ -0,0 +1,142 @@
+
+2. **Scalability**:
+   - The game architecture should handle an increase in the number of players
+     and game sessions efficiently.
+
+3. **Security**:
+   - Player data must be protected against unauthorized access and breaches.
+   - The system must include secure authentication and authorization mechanisms.
+
+4. **Reliability**:
+   - The game should handle network disruptions gracefully, with mechanisms for
+     reconnection and state preservation.
+
+5. **Usability**:
+   - The user interface should be intuitive and easy to navigate.
+   - The game should provide clear feedback to players for their actions.
+
+6. **Maintainability**:
+   - The codebase should be modular and well-documented to facilitate easy
+     maintenance and updates.
+
+### Optional features
+
+1. **Controller support**: The game should optionally support game controllers.
+
+## Architecture overview
+
+### System architecture
+
+Project OMG game architecture is based on a monolithic design, where all
+components are integrated into a single codebase. This approach simplifies
+development and deployment, making it suitable for the project's initial scope.
+The architecture consists of three main components:
+
+- [Client application][client_app_doc]
+- [Game server][game_server_doc]
+- [Database][database_doc]
+
+### Functional requirements coverage by components
+
+| Functional requirement | Client application | Game server | Database |
+|------------------------|--------------------|-------------|----------|
+| Multiplayer support    |                    |      ✓      |          |
+| Top-down 2D view       |         ✓          |             |          |
+| Character lock on camera |       ✓          |             |          |
+| Match-based gameplay   |         ✓          |      ✓      |          |
+| Character creation     |         ✓          |      ✓      |     ✓    |
+| Skill synergy          |         ✓          |      ✓      |          |
+| PvPvE gameplay         |                    |      ✓      |          |
+| Intelligent bots       |                    |      ✓      |          |
+| Interactive environment |        ✓          |             |          |
+| End-game statistics    |         ✓          |      ✓      |     ✓    |
+| Monitoring statistics throughout the game | ✓ |      ✓      |     ✓  |
+
+### High-level architecture diagram
+
+![High-Level Architecture](figures/hld_arch.png)
+
+### Component overview
+
+#### [Client application][client_app_doc]
+
+- **Game UI**: The user interface for the game, including menus, in-game HUD,
+  and end-game statistics.
+- **Network manager**: Handles communication with the Game Server, ensuring
+  real-time updates.
+- **Game logic**: Processes user inputs, updates the game state locally, and
+  synchronizes with the server.
+
+[client_app_doc]: hld_client_app.md
+
+#### [Game server][game_server_doc]
+
+- **Matchmaker**: Manages player matchmaking, pairing players based on skill
```

Comment:
Doesn't this item contradict the idea of "no saved progression"?

Reply:
I believe progression is in terms of in-game progression. It does not include out-of-game rankings or elo.

Reply:
"pairing players based on skill level" sounds as if we store the player progression data(e.g. level, ranking, elo etc).

Reply:
I removed that phrase to unify the understanding. Indeed we won't plan on ranking elo or another type of data stored.